crates/containerd-shim-wasm/src/sandbox/cgroups.rs

```diff
@@ -0,0 +1,674 @@
+    // for details on converting the oci spec (which is heavily v1 focussed) to v2.
+    // Also https://github.com/containers/crun/blob/2497b9bb03623838d37a1587087f1ad3d6ff28ec/crun.1.md#cgroup-v2
+    fn apply(&self, res: Option<Resources>) -> Result<()> {
+        if res.is_none() {
+            return Ok(());
+        }
+        let res = res.unwrap();
+
+        if let Some(cpu) = res.cpu() {
+            if let Some(shares) = cpu.shares() {
+                let s = 1 + ((shares - 2) * 9999) / 262142;
+                ensure_write_file(self.get_file("cpu.weight"), &format!("{}", s))?;
+            }
+            let mut max = "max".to_string();
+            if let Some(quota) = cpu.quota() {
+                max = format!("{}", quota);
+            }
+            if let Some(period) = cpu.period() {
+                ensure_write_file(self.get_file("cpu.max"), &format!("{} {}", max, period))?;
+            } else {
+                ensure_write_file(self.get_file("cpu.max"), max.as_str())?;
+            }
+
+            // no realtime support
+        }
+
+        if let Some(mem) = res.memory() {
+            if let Some(limit) = mem.limit() {
+                ensure_write_file(self.get_file("memory.max"), &format!("{}", limit))?;
+
+                if let Some(swap) = mem.swap() {
+                    // OCI spec expects swap to be memory+swap (because that's how cgroup v1 does things)
+                    // V2 expects just the swap limit, so we need to subtract the swap limit (again, memory+swap) from the memory limit to get the swap total.
+                    ensure_write_file(
+                        self.get_file("memory.swap.max"),
+                        &format!("{}", limit - swap),
+                    )?;
+                }
+            }
+
+            if let Some(reservation) = mem.reservation() {
+                ensure_write_file(self.get_file("memory.low"), &format!("{}", reservation))?;
+            }
+
+            if let Some(oom_kill_disable) = mem.disable_oom_killer() {
+                ensure_write_file(
+                    self.get_file("memory.oom.group"),
+                    if oom_kill_disable { "1" } else { "0" },
+                )?;
+            }
+        }
+
+        if let Some(pids) = res.pids() {
+            ensure_write_file(self.get_file("pids.max"), &format!("{}", pids.limit()))?;
+        }
+
+        if let Some(hugepage) = res.hugepage_limits() {
+            for limit in hugepage {
+                let path = self.get_file(&format!("hugetlb.{}.max", limit.page_size()));
+                ensure_write_file(path, &format!("{}", limit.limit()))?;
+            }
+        }
+
+        if let Some(blkio) = res.block_io() {
+            if let Some(weight) = blkio.weight() {
+                ensure_write_file(
+                    self.get_file("io.weight"),
+                    &format!("{}", (weight - 10) * 9999 / 990),
```

Comment:
https://github.com/containerd/cgroups/blob/724eb82fe759f3b3b9c5f07d22d2fab93467dc56/v2/utils.go#L209

Was it a mistake? Missing 1 plus here

```suggestion
                    &format!("{}", 1 + (weight - 10) * 9999 / 990),
```

Reply:
Nooooo, off by one!